arduino/mono_encoder/mono_encoder.ino

```diff
@@ -0,0 +1,318 @@
+ * will be used to maintain the tick count.  The code 
+ * implements a robust debouncing technique.  
+ * The advantage of this mode is the debounce logic;
+ * it can robustly count ticks even from a noisy
+ * mechanical encoder like a KY-040 rotary encoder.
+ * The disadvantage is that this cannot handle high rates
+ * of ticks as would be delivered by a high resolution optical
+ * encoder.
+ *
+ * The method debounces noisy inputs,
+ * looking for stable transitions from
+ * open-to-closed and closed-to-open by polling
+ * the encoder pin and building a 16 bit value.
+ * We start by looking for a stable closed-to-open
+ * transition, which will be a 1 followed by 15 zeroes.
+ * Then we change to look for the open-to-closed transition,
+ * which is a zero reading followed by 15 ones.
+ *
+ * Given a polling delay of 100 microseconds, the
+ * minimum time for a stable reading is
+ * 100 microseconds * 16 bits = 1.6 milliseconds.
+ * That then would allow for a maximum read rate of
+ * 1000 / 1.6 = 625 ticks per second.
+ * For high resolution encoders, that may be too slow.
+ * In that case, reduce the POLL_DELAY_MICROS to
+ * suit your encoder and use case.
+ */
+#include <Arduino.h>
+
+#define ENCODER_PIN   (2)          // input pin for first encoder
+#define ENCODER_2_PIN (3)          // input pin for second encoder
+#define POLL_DELAY_MICROS (100UL)  // microseconds between polls
+#define USE_ENCODER_INTERRUPTS (1) // if defined then use interrupts to read ticks
+                                   // otherwise use polling.
+
+//
+// state for an encoder pin
+//
+struct EncoderState {
+  int pin;                     // gpio pin number
+  volatile long ticks;         // current tick count; do NOT read this directly, use readEncoders()
+  uint16_t readings;           // history of last 16 readings
+  uint16_t transition;         // value of last 16 readings for next stable state transition 
+  unsigned long pollAtMicros;  // time of next poll
+};
+
+// list of encoders
+static EncoderState encoders[] = {
+  {ENCODER_PIN, 0L, 0, 0x8000, 0UL},
+  #ifdef ENCODER_2_PIN
+    {ENCODER_2_PIN, 0L, 0, 0x8000, 0UL},
+  #endif
+};
+
+#define ENCODER_COUNT (sizeof(encoders) / sizeof(EncoderState))
+
+boolean continuous = false; // true to send continuously, false to only send on demand
+unsigned long sendAtMs = 0; // next send time in continuous mode
+unsigned long delayMs = 0;  // time between sends in continuous mode
+
+#ifdef USE_ENCODER_INTERRUPTS
+  typedef void (*gpio_isr_type)();
+
+  //
+  // count the interrupts
+  //
+  void encode_0() {
+    ++encoders[0].ticks;
+  }
+
+  #ifdef ENCODER_2_PIN
+    void encode_1() {
+      ++encoders[1].ticks;
+    }
+  #endif
+
+  gpio_isr_type _isr_routines[ENCODER_COUNT] = {
+      encode_0,
+      #ifdef ENCODER_2_PIN
+        encode_1
+      #endif
+  };
+#endif
+
+
+
+//
+// Poll the encoder and increment ticks
+// on stable transitions.
+//
+void pollEncoder(EncoderState &encoder) {
```

Comment:
You could make this a member function.

Reply:
I am using structs rather than classes; this is structured like C code more than C++.  I think that is typical of sketches in Arduino (although libraries typically use C++ classes for encapsulation).

Reply:
Structs are like classes, just you don't have public / private / protected, i.e. access rights as you say, and everything is public. But you do have member functions, polymorphism, etc. The Arduino compiler (I think it's gcc) understands C++. If you don't use polymorphism, i.e. vtable, then you shouldn't have any overhead in compiled code. I agree that I have seen people frequently writing complicated, antique C-style code for Arduino, instead of modern OO C++ code. Which I found strange.